### PR TITLE
Add support for recording multiple CPEs in cyclonedx

### LINKED ISSUE #818: Add support for multiple CPEs in CycloneDX
**What would you like to be added**:

Currently syft only records the first CPE in its outputs for CycloneDX. While looking at the CycloneDX spec and the SPDX spec, I remembered that similar to SPDX, cyclonedx supports adding external references of type "advisory" in a URI format. CPE, coincidentally can be stored either in a string format or a URI format. Syft already supports parsing URL formatted CPEs, but it doesn't currently export CPEs in URI format. If we combine these two properties, we should be able to save all the CPE information in the CycloneDX format as well while still storing the primary or what we assume is the canonical CPE in the CycloneDX CPE field which only allows for 1 CPE. 

**Why is this needed**:

While there can only be 1 canonical CPE for any package, it is impossible for scanners to accurately determine this. While syft puts the most specific CPE in the CycloneDX CPE field, there are times where CPEs have alternative vendors etc. that we may want to capture. For these use cases, we should allow syft to store all its generated CPEs in the output cyclonedx bom.

### PR DESCRIPTION
**Pull request by samj1912**
Fixes #818 

For reasoning and logic behind the CPE serialization decision please take a look at the issue. We still store our most specific CPE in the cyclonedx CPE field but for all the other CPEs, we store them as external references.

### CODE REVIEW

File: syft/pkg/cpe_test.go [hunk 1]
```diff
@@ -105,6 +109,9 @@ func Test_CPEParser(t *testing.T) {
 			assert.Equal(t, c1, test.WFN)
 			assert.Equal(t, c2, test.WFN)
 			assert.Equal(t, CPEString(test.WFN), test.CPEString)
+			assert.Equal(t, CPEURI(c2), test.CPEUrl)
+			assert.Equal(t, CPEURI(c1), test.CPEUrl)
+			assert.Equal(t, must(NewCPE(CPEURI(c1))), test.WFN)
```
Comment (samj1912) [inline syft/pkg/cpe_test.go:114]: This ensures that we are also making our URL generation also go through the same rigorous CPE encoding/decoding test that our CPE strings go through.

File: syft/pkg/cpe.go [hunk 2]
```diff
@@ -99,6 +99,22 @@ func CPEString(c CPE) string {
...
+	output.Vendor = sanitize(c.Vendor)
+	output.Product = sanitize(c.Product)
+	output.Language = sanitize(c.Language)
+	output.Version = sanitize(c.Version)
+	output.TargetSW = sanitize(c.TargetSW)
+	output.Part = sanitize(c.Part)
+	output.Edition = sanitize(c.Edition)
+	output.Other = sanitize(c.Other)
+	output.SWEdition = sanitize(c.SWEdition)
+	output.TargetHW = sanitize(c.TargetHW)
```
Comment (samj1912) [inline syft/pkg/cpe.go:113]: This mimics our custom CPEString logic above to handle special characters in the CPE.

**Comment by samj1912** [comment]
cc: @coderpatros & @stevespringett to validate if this seems like a reasonable usage of the cyclonedx spec. to get around accurate CPE determination limitation that scanning tools face. Detailed reasoning at https://github.com/anchore/syft/issues/818

**Comment by stevespringett** [comment]
I think using properties is more appropriate. External references in CycloneDX are typically URIs to other resources. This varies a lot from SPDX which applies the external reference name to mean potential alternative identifiers.

We have a formal property taxonomy as well. https://github.com/CycloneDX/cyclonedx-property-taxonomy

We could register `anchore` as a top-level namespace if desired?

**Comment by samj1912** [comment]
Thanks for the feedback, I have opened a PR to register a `syft` namespace since we are also using cyclonedx properties for storing other fields. The PR should first be reviewed by syft maintainers before any further action is taken -> https://github.com/CycloneDX/cyclonedx-property-taxonomy/pull/8


cc: @luhring, @wagoodman, @spiffcs, @kzantow please take a look. My reasoning for choosing `syft` over `anchore` was in case there was ever some movement in the future on https://github.com/ossf/tac/issues/70. The choice of a namespace like syft ensures that the namespace is easily identifiable and closely associated to syft while still being impervious to such changes.

**Comment by samj1912** [comment]
Separately it looks like there is an spdx taxonomy in work as well which contains the CPE field at https://github.com/CycloneDX/cyclonedx-property-taxonomy/pull/7 which might also be suitable for storing this and other SPDX fields. This might also be useful with reference to #563, #723 and #737

**Comment by kzantow** [comment]
@samj1912 since https://github.com/anchore/syft/pull/849 has been merged, which outputs `syft:` namespace'd CPEs, I'm going to close this one; please reopen if you disagree!

**Pull request closed without merging**